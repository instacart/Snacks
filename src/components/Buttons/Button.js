import React from 'react'
import PropTypes from 'prop-types'
import Radium from 'radium'
import Icon from '../Icon/Icon'
import withTheme from '../../styles/themer/withTheme'
import { themePropTypes } from '../../styles/themer/utils'
import { spacing, colors } from '../../styles'
import { darken } from '../../utils'

const noop = () => {} // eslint-disable-line no-empty-function

const baseStyles = {
  touchAction: 'manipulation',
  cursor: 'pointer',
  border: '1px solid transparent',
  borderTopRightRadius: 4,
  borderBottomRightRadius: 4,
  borderBottomLeftRadius: 4,
  borderTopLeftRadius: 4,
  fontWeight: 600,
  whiteSpace: 'nowrap',
  userSelect: 'none',
  WebkitFontSmoothing: 'antialiased',

  // Reset unusual Firefox-on-Android default style;
  // see https://github.com/necolas/normalize.css/issues/214
  backgroundImage: 'none',

  // Takes care of the heavier default font-weights in firefox
  MozOsxFontSmoothing: 'grayscale',

  // Ensures any icons stay vertically centered
  display: 'inline-flex',
  alignItems: 'center'
}

const sizeStyles = {
  tiny: {
    ...spacing.PADDING_X_XS,
    height: '24px',
    fontSize: '11px',
    textTransform: 'uppercase'
  },

  small: {
    ...spacing.PADDING_X_XS,
    fontSize: '14px',
    height: '32px'
  },

  standard: {
    ...spacing.PADDING_X_SM,
    fontSize: '16px',
    height: '40px'
  },

  large: {
    ...spacing.PADDING_X_MD,
    fontSize: '18px',
    height: '48px'
  }
}

const linkStyles = {
  display: 'flex',
  alignItems: 'center',
  boxSizing: 'border-box'
}

const getSnacksStyles = props => {
  const { action, actionHover, primaryBackground } = props.snacksTheme.colors
  const actionActive = darken(actionHover, 3)

  const disabled = {
    backgroundColor: colors.GRAY_74,
    color: colors.WHITE,
    cursor: 'not-allowed'
  }

  const activeAndFocus = {
    primary: {
      base: {
        backgroundColor: actionActive
      },
      inverted: {
        backgroundColor: primaryBackground,
        opacity: '0.8'
      }
    },
    secondary: {
      base: {
        color: actionActive,
        border: `1px solid ${actionActive}`
      },
      inverted: {
        color: primaryBackground,
        border: `1px solid ${primaryBackground}`,
        opacity: '0.8'
      }
    },
    flat: {
      color: actionActive,
    },
    coupon: {
      border: `1px dashed ${colors.RED_700}`,
      color: colors.RED_700
    }
  }

  return {
    primary: {
      base: {
        backgroundColor: action,
        color: primaryBackground,

        ':hover': {
          backgroundColor: actionHover
        },
        ':active': activeAndFocus.primary.base,
        ':focus': activeAndFocus.primary.base
      },

      inverted: {
        backgroundColor: primaryBackground,
        border: `1px solid ${action}`,
        color: action,

        ':hover': {
          backgroundColor: primaryBackground,
          opacity: '0.9'
        },
        ':active': activeAndFocus.primary.inverted,
        ':focus': activeAndFocus.primary.inverted
      },

      disabled
    },

    secondary: {
      base: {
        backgroundColor: 'transparent',
        color: action,
        border: `1px solid ${action}`,

        ':hover': {
          color: actionHover,
          border: `1px solid ${actionHover}`
        },
        ':active': activeAndFocus.secondary.base,
        ':focus': activeAndFocus.secondary.base
      },

      inverted: {
        color: primaryBackground,
        border: `1px solid ${primaryBackground}`,

        ':hover': {
          color: primaryBackground,
          border: `1px solid ${primaryBackground}`,
          opacity: '0.9'
        },
        ':active': activeAndFocus.secondary.inverted,
        ':focus': activeAndFocus.secondary.inverted
      },

      disabled
    },

    flat: {
      base: {
        backgroundColor: 'transparent',
        color: action,

        ':hover': {
          color: actionHover,
        },
        ':active': activeAndFocus.flat,
        ':focus': activeAndFocus.flat
      },

      disabled: {
        backgroundColor: 'transparent',
        color: colors.GRAY_74,
        cursor: 'not-allowed'
      }
    },

    coupon: {
      base: {
        backgroundColor: colors.WHITE,
        color: colors.RED_500,
        border: `1px dashed ${colors.RED_500}`,

        ':hover': {
          border: `1px dashed ${colors.RED_600}`,
          color: colors.RED_600
        },
        ':active': activeAndFocus.coupon,
        ':focus': activeAndFocus.coupon
      },

      disabled
    }
  }
}

const Button = props => {
  const snacksStyles = getSnacksStyles(props)

  const ElementType = props.href ? 'a' : 'button'

  const finalProps = {
    disabled: props.disabled,
    tabIndex: props.tabIndex,
    type: props.type,
    style: [
      baseStyles,
      sizeStyles[props.size],
      snacksStyles[props.snacksStyle][props.disabled ? 'disabled' : 'base'],
      props.inverted && snacksStyles[props.snacksStyle].inverted,
      ElementType === 'a' && linkStyles,
      props.style
    ],
    onClick: e => {
      if (props.disabled) {
        e.preventDefault()
        return
      }

      props.onClick(e, props)
    },
    onMouseDown: e => {
      if (props.disabled) {
        e.preventDefault()
        return
      }

      props.onMouseDown(e, props)
    },
    ...props.elementAttributes
  }
  if (props.href) {
    finalProps.href = props.href
  }

  const icon =
    typeof props.icon === 'string' ? <Icon name={props.icon} /> : props.icon

  if (icon && props.iconPosition === 'left') {
    return (
      <ElementType {...finalProps}>
        {icon}
        <span style={{ ...spacing.MARGIN_RIGHT_XS }} />
        {props.children}
      </ElementType>
    )
  }

  if (icon && props.iconPosition === 'right') {
    return (
      <ElementType {...finalProps}>
        {props.children}
        <span style={{ ...spacing.MARGIN_LEFT_XS }} />
        {icon}
      </ElementType>
    )
  }

  return <ElementType {...finalProps}>{props.children}</ElementType>
}

Button.propTypes = {
  /** Size of the button. */
  size: PropTypes.oneOf(['tiny', 'small', 'standard', 'large']),

  /** Snacks button variants. */
  snacksStyle: PropTypes.oneOf(['primary', 'secondary', 'flat', 'coupon']),

  /** Optional style overrides. */
  style: PropTypes.object,

  /** Whether or not the button is disabled. */
  disabled: PropTypes.bool,

  /** Sets the HTML type attribute on the element. */
  type: PropTypes.oneOf(['button', 'reset', 'submit']),

  /**
    An optional icon. Can be a an icon name
    or a Snacks `Icon` component.
  */
  icon: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),

  /** Where the icon is positioned relative to primary text content. */
  iconPosition: PropTypes.oneOf(['left', 'right']),

  /** Callback that fires when the button is clicked. */
  onClick: PropTypes.func,

  /** Callback that fires on mouse down. */
  onMouseDown: PropTypes.func,

  /** Reverses colors. Use for rendering buttons on dark backgrounds. */
  inverted: PropTypes.bool,

  /** Sets the HTML `tabindex` attribute. */
  tabIndex: PropTypes.number,

  /**
    If href is provided, `Button` will render as an anchor
    tag with the supplied href value.
  */
  href: PropTypes.string,

  /** Text content for the button. */
  children: PropTypes.node,

  /** Any additonal props to add to the element (e.g. data attributes). */
  elementAttributes: PropTypes.object,

  /** Snacks theme attributes provided by `Themer` */
  snacksTheme: themePropTypes
}

Button.defaultProps = {
  size: 'standard',
  snacksStyle: 'primary',
  disabled: false,
  type: 'button',
  iconPosition: 'left',
  onClick: noop,
  onMouseDown: noop,
  inverted: false,
  elementAttributes: {}
}

export default withTheme(Radium(Button))
