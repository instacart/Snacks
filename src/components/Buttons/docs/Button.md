This is our Button.
There are many like it, but this one is ours.

Buttons guide our users to take a certain action, like 'Add to Cart' or 'Checkout'.
The call to action should start with an active verb, and the style choice of the button should reflect the hierarchy and importance of the action.

### Primary
Primary buttons are for, you guessed it, the primary action.
Don't use these lightly.
Overuse on a single screen could hinder your main objective rather than help it.

```js
<div style={{display: 'flex'}}>
  <div style={{marginRight: '24px'}}>
    <div style={{marginBottom: '8px'}}>Tiny</div>
    <Button snacksStyle="primary" size="tiny">
      Primary Button
    </Button>
  </div>
  <div style={{marginRight: '24px'}}>
    <div style={{marginBottom: '8px'}}>Small</div>
    <Button snacksStyle="primary" size="small">
      Primary Button
    </Button>
  </div>
  <div style={{marginRight: '24px'}}>
    <div style={{marginBottom: '8px'}}>Standard</div>
    <Button snacksStyle="primary" size="standard">
      Primary Button
    </Button>
  </div>
  <div>
    <div style={{marginBottom: '8px'}}>Large</div>
    <Button snacksStyle="primary" size="large">
      Primary Button
    </Button>
  </div>
</div>
```

### Secondary
Secondary buttons are for your second most important action.
A good use is as a cancel button when asking the user a question.

```js
<div style={{display: 'flex'}}>
  <div style={{marginRight: '24px'}}>
    <div style={{marginBottom: '8px'}}>Tiny</div>
    <Button snacksStyle="secondary" size="tiny">
      Secondary Button
    </Button>
  </div>
  <div style={{marginRight: '24px'}}>
    <div style={{marginBottom: '8px'}}>Small</div>
    <Button snacksStyle="secondary" size="small">
      Secondary Button
    </Button>
  </div>
  <div style={{marginRight: '24px'}}>
    <div style={{marginBottom: '8px'}}>Standard</div>
    <Button snacksStyle="secondary" size="standard">
      Secondary Button
    </Button>
  </div>
  <div>
    <div style={{marginBottom: '8px'}}>Large</div>
    <Button snacksStyle="secondary" size="large">
      Secondary Button
    </Button>
  </div>
</div>
```

### Flat
The flat style can be used when you want a button that looks like a link.
A typical solution is to use an anchor tag to do this, however this approach
has issues such as not having consistent padding to a button and there are
accessibility concerns (links should navigate, buttons should cause actions).

```js
<div style={{display: 'flex'}}>
  <div style={{marginRight: '24px'}}>
    <div style={{marginBottom: '8px'}}>Tiny</div>
    <Button snacksStyle="flat" size="tiny">
      Flat Button
    </Button>
  </div>
  <div style={{marginRight: '24px'}}>
    <div style={{marginBottom: '8px'}}>Small</div>
    <Button snacksStyle="flat" size="small">
      Flat Button
    </Button>
  </div>
  <div style={{marginRight: '24px'}}>
    <div style={{marginBottom: '8px'}}>Standard</div>
    <Button snacksStyle="flat" size="standard">
      Flat Button
    </Button>
  </div>
  <div>
    <div style={{marginBottom: '8px'}}>Large</div>
    <Button snacksStyle="flat" size="large">
      Flat Button
    </Button>
  </div>
</div>
```

### Coupon

Note that coupon buttons aren't affected by a Snacks theme.

```js
<div style={{display: 'flex'}}>
  <div style={{marginRight: '24px'}}>
    <div style={{marginBottom: '8px'}}>Tiny</div>
    <Button snacksStyle="coupon" size="tiny">
      Coupon Button
    </Button>
  </div>
  <div style={{marginRight: '24px'}}>
    <div style={{marginBottom: '8px'}}>Small</div>
    <Button snacksStyle="coupon" size="small">
      Coupon Button
    </Button>
  </div>
  <div style={{marginRight: '24px'}}>
    <div style={{marginBottom: '8px'}}>Standard</div>
    <Button snacksStyle="coupon" size="standard">
      Coupon Button
    </Button>
  </div>
  <div>
    <div style={{marginBottom: '8px'}}>Large</div>
    <Button snacksStyle="coupon" size="large">
      Coupon Button
    </Button>
  </div>
</div>
```

### Disabled

Disabled buttons are to show the user there is a future action they can take, but they must first accomplish another task.

```js
<div>
  <span style={{marginRight: '24px'}}>
    <Button disabled snacksStyle="primary" size="tiny">
      Disabled Button
    </Button>
  </span>
  <span style={{marginRight: '24px'}}>
    <Button disabled snacksStyle="primary" size="small">
      Disabled Button
    </Button>
  </span>
  <span style={{marginRight: '24px'}}>
    <Button disabled snacksStyle="primary" size="standard">
      Disabled Button
    </Button>
  </span>
  <span>
    <Button disabled snacksStyle="primary" size="large">
      Disabled Button
    </Button>
  </span>
</div>
```

### Inverted colors
Button colors can be inverted for darker backgrounds.

```js
<div style={{padding: '24px', backgroundColor: '#43B02A'}}>
  <span style={{marginRight: '24px'}}>
    <Button inverted snacksStyle="primary" size="standard">
      Primary Button
    </Button>
  </span>
  <span style={{marginRight: '24px'}}>
    <Button inverted snacksStyle="secondary" size="standard">
      Secondary Button
    </Button>
  </span>
  <span>
    <Button disabled snacksStyle="primary" size="standard">
      Disabled Button
    </Button>
  </span>
</div>
```

### With icons
Buttons can be rendered with an icon to the left or right of the primary content.

```js
<div style={{padding: '24px'}}>
  <span style={{marginRight: '24px'}}>
    <Button iconPosition="right" icon={<Icon name="arrowRightSmallBold" />} snacksStyle="secondary" size="standard">
      View 24 more
    </Button>
  </span>
  <span style={{marginRight: '24px'}}>
    <Button icon="cart" snacksStyle="primary" size="standard">
      Cart
    </Button>
  </span>
  <span style={{marginRight: '24px'}}>
    <Button icon={<Icon name="deals" />} snacksStyle="coupon" size="small">
      Save $5.00
    </Button>
  </span>
</div>
```
