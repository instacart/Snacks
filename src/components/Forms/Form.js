import React     from 'react'
import PropTypes from 'prop-types'
import _         from 'underscore'

class Form extends React.Component {
  static propTypes = {
    /** Form html chilren */
    children    : PropTypes.node,
    /** HTML form attributes */
    formProps   : PropTypes.shape({}),
    /** onSubmit callback will pass in model as parameter */
    onSubmit    : PropTypes.func,
    /** errors from server mapped to model names. Will attach serverErrors styling to FormComppnents */
    serverErrors: PropTypes.shape({})
  }

  static childContextTypes = {
    ICFormable: PropTypes.object
  }

  constructor() {
    super()
    this.state          = { serverErrors: null }
    this.model          = {}
    this.formComponents = {}
  }

  getChildContext() {
    return {
      ICFormable: {
        registerComponent  : this.registerComponent,
        unregisterComponent: this.unregisterComponent
      }
    }
  }

  componentDidMount() {
    this.updateModel()
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.serverErrors && !_.isEqual(this.state.serverErrors, nextProps.serverErrors)) {
      this.setState({serverErrors: nextProps.serverErrors}, () => {
        this.setErrorsOnFormComponents(this.state.serverErrors)
      })
    }
  }

  registerComponent = (component) => {
    this.formComponents[component.props.name] = component
  }

  unregisterComponent = (component) => {
    delete this.formComponents[component.props.name]
    delete this.model[component.props.name]
  }

  setErrorsOnFormComponents(serverErrors) {
    Object.keys(serverErrors).forEach( (name) => {
      const component = this.formComponents[name]
      component.setState({
        isValid: false,
        serverError: serverErrors[name]
      })
    })
  }

  formIsValid() {
    const components = Object.values(this.formComponents)
    return components.every( component => component.validate() )
  }

  updateModel() {
    Object.keys(this.formComponents).forEach((name) => {
      const component = this.formComponents[name]
      if (component.props.disabled) {
        delete this.model[name]
      } else {
        this.model[name] = component.getValue()
      }
    })
  }

  handleSubmit = (e) => {
    e && e.preventDefault()
    this.updateModel()

    if (this.formIsValid()) {
      this.setState({serverErrors: null}, () => {
        this.props.onSubmit && this.props.onSubmit(this.model)
      })
    }
  }

  render() {
    const {
      children,
      formProps
    } = this.props

    return (
      <form {...formProps} onSubmit={this.handleSubmit}>
        {children}
      </form>
    )
  }
}

export default Form
