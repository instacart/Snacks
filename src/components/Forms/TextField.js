import React from 'react'
import PropTypes from 'prop-types'
import { colors } from '../../styles'
import FormComponent from './FormComponent'
import ValidationError from './ValidationError'
import FloatingLabel from './FloatingLabel'
import TextFieldHint from './TextFieldHint'
import ServerError from './ServerError'
import HelperText from './HelperText'
import withTheme from '../../styles/themer/withTheme'
import { themePropTypes } from '../../styles/themer/utils'
import spacing from '../../styles/spacing'

const styles = {
  wrapper: {
    cursor: 'auto',
    display: 'inline-block',
    position: 'relative',
    width: '343px',
  },
  inputContainer: {
    borderRadius: '4px',
    position: 'relative',
  },
  input: {
    backgroundColor: '#FFF',
    border: `solid 1px ${colors.GRAY_74}`,
    borderRadius: '4px',
    boxSizing: 'border-box',
    color: colors.GRAY_20,
    fontSize: '16px',
    height: '56px',
    marginTop: 0,
    marginRight: 0,
    marginBottom: 0,
    marginLeft: 0,
    paddingTop: '25px',
    ...spacing.PADDING_X_XS,
    paddingBottom: spacing.XS,
    position: 'relative',
    width: '100%',
    WebkitOpacity: 1,
    WebkitTapHighlightColor: 'rgba(0,0,0,0)',
  },
  inputDisabled: {
    border: `1px dashed ${colors.GRAY_74}`,
    backgroundColor: colors.GRAY_93,
    color: colors.GRAY_46,
    cursor: 'not-allowed',
  },
  inputError: {
    border: `1px solid ${colors.RED_700}`,
    backgroundColor: '#FDE6EB',
  },
  fullWidth: {
    width: '100%',
  },
  halfWidth: {
    width: '162px',
  },
}

const getSnackStyles = snacksTheme => {
  const { action } = snacksTheme.colors
  return {
    highlight: {
      border: `1px solid ${action}`,
    },
  }
}

@withTheme
@FormComponent
class TextField extends React.Component {
  static propTypes = {
    className: PropTypes.string,
    /** Name of the field */
    name: PropTypes.string.isRequired,
    /** HTML autocomplete attribute */
    autoComplete: PropTypes.string,
    /** DefaultValue for non controlled component */
    defaultValue: PropTypes.any,
    /** Disable the text field */
    disabled: PropTypes.bool,
    /** Text of label that will animate when TextField is focused */
    floatingLabelText: PropTypes.string,
    /** Sets width to 100% */
    fullWidth: PropTypes.bool,
    /** Sets width to 162px */
    halfWidth: PropTypes.bool,
    /** FormComponent error for validation */
    hasError: PropTypes.bool,
    /** Helper text will show up in bottom right corner below TextField */
    helperText: PropTypes.string,
    /** Hint text will show up when input is focused and there is no value */
    hintText: PropTypes.string,
    /** Uniq id for input */
    id: PropTypes.string,
    /** className for input */
    inputClassName: PropTypes.string,
    /** Style for input */
    inputStyle: PropTypes.object,
    /** Set by FormComponent by default.   */
    isValid: PropTypes.bool,
    /** onFocus callback */
    onFocus: PropTypes.func,
    /** onChange callback */
    onChange: PropTypes.func,
    /** onBlur callback */
    onBlur: PropTypes.func,
    /** onKeyDown callback */
    onKeyDown: PropTypes.func,
    /** Mark the field as required.  */
    required: PropTypes.bool,
    /** Error from server to show ServerError message */
    serverError: PropTypes.string,
    /**
     * Optional style overrides merged into emotion css
     *
     * @deprecated
     * This prop exists for backwards compatibility and will be
     * removed in a future version
     */
    style: PropTypes.object,
    /** Input type ie. 'text', 'email', password, etc..  */
    type: PropTypes.string,
    /** Text to show for validation error  */
    validationErrorText: PropTypes.string,
    /** Value will make TextField a controlled component  */
    value: PropTypes.string,
    /** Snacks theme attributes provided by `Themer` */
    snacksTheme: themePropTypes,
    /** Any additonal props to add to the element (e.g. data attributes) */
    elementAttributes: PropTypes.object,
  }

  static defaultProps = {
    autoComplete: 'on',
    disabled: false,
    type: 'text',
    defaultValue: null,
    onKeyDown: () => {}, // eslint-disable-line no-empty-function
  }

  state = {
    hasValue: this.props.defaultValue !== null || Boolean(this.props.value),
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.disabled && !this.props.disabled) {
      this.setState({ isFocused: false })
    }

    if (!this.state.hasValue && nextProps.value) {
      this.setState({ hasValue: true })
    }
  }

  getValue = () => {
    if (!this.input) {
      return null
    }

    return this.input.value
  }

  handleInputChange = e => {
    const { onChange } = this.props
    const { hasValue } = this.state
    const { value } = e.target
    // Limit setState call to only when hasValue changes
    if (value && !hasValue) {
      this.setState({ hasValue: true })
    } else if (!value && hasValue) {
      this.setState({ hasValue: false })
    }

    onChange && onChange(e, value)
  }

  handleInputFocus = e => {
    this.setState({ isFocused: true })
    this.props.onFocus && this.props.onFocus(e)
  }

  handleInputBlur = e => {
    this.setState({ isFocused: false })
    this.props.onBlur && this.props.onBlur(e)
  }

  handleKeyDown = e => {
    this.props.onKeyDown(e)
  }

  triggerFocus = () => this.input.focus()

  render() {
    const {
      className,
      floatingLabelText,
      defaultValue,
      disabled,
      fullWidth,
      halfWidth,
      hasError,
      hintText,
      id,
      inputClassName,
      inputStyle,
      isValid,
      name,
      required,
      serverError,
      type,
      validationErrorText,
      style,
      value,
      helperText,
      autoComplete,
      snacksTheme,
      elementAttributes,
    } = this.props

    const { hasValue, isFocused } = this.state

    const snacksStyles = getSnackStyles(snacksTheme)

    const inputId = id
    const showHintText = hintText && !hasValue && isFocused

    return (
      <div
        className={className}
        css={[styles.wrapper, fullWidth && styles.fullWidth, halfWidth && styles.halfWidth, style]}
      >
        {serverError && !disabled && !isValid && <ServerError text={serverError} />}

        <div css={styles.inputContainer}>
          {floatingLabelText && (
            <FloatingLabel
              text={floatingLabelText}
              float={isFocused || hasValue}
              disabled={disabled}
              isActive={isFocused}
              hasError={hasError}
              htmlFor={inputId}
              snacksTheme={snacksTheme}
            />
          )}

          {hintText && (
            <TextFieldHint
              inputId={`hint_${inputId}`}
              text={hintText}
              show={showHintText}
              disabled={disabled}
            />
          )}

          <input
            value={value}
            id={inputId}
            ref={node => {
              this.input = node
            }}
            defaultValue={value !== undefined ? undefined : defaultValue}
            disabled={disabled}
            name={name}
            type={type}
            aria-required={required}
            aria-invalid={hasError}
            aria-describedby={[
              hasError ? `error_${inputId}` : null,
              hintText ? `hint_${inputId}` : null,
            ]
              .filter(Boolean)
              .join(' ')}
            className={inputClassName}
            css={[
              styles.input,
              inputStyle,
              disabled && styles.inputDisabled,
              !disabled && hasError && styles.inputError,
              isFocused && !hasError && snacksStyles.highlight,
            ]}
            onBlur={this.handleInputBlur}
            onChange={this.handleInputChange}
            onFocus={this.handleInputFocus}
            onKeyDown={this.handleKeyDown}
            autoComplete={autoComplete}
            placeholder=""
            {...elementAttributes}
          />
        </div>

        <ValidationError
          text={validationErrorText}
          show={!disabled && !isValid && !serverError}
          inputId={inputId}
        />

        <HelperText helperText={helperText} />
      </div>
    )
  }
}

export default TextField
