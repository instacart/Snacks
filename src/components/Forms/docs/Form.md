
```js
<Form
  onSubmit={(model) => ( console.log(model) ) }
  serverErrors={{}}
  formProps={{}}
>
  <div style={{width: '335px'}}>
    <div style={{marginBottom: '10px'}}>
      <TextField
        name="email"
        type="email"
        floatingLabelText="Email"
        hintText="Enter your email address"
        validations={{isEmail: null, isLength: {min: 3, max: 50}}}
        validationErrorText="Please enter a valid email"
        fullWidth
        required
      />
    </div>

    <div>
      <TextField
        name="password"
        type="password"
        floatingLabelText="Password"
        hintText="Enter a secure password"
        validations={{isLength: {min: 6}}}
        validationErrorText="Password must be at least 6 characters"
        fullWidth
        required
      />
    </div>

    <div style={{marginTop: '10px'}}>
      <Button type="submit">
        Submit
      </Button>
    </div>
  </div>
</Form>

```
