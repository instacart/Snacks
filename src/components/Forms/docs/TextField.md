TextField Example:

```js
<div>
  <TextField
    name="test"
    type="email"
    floatingLabelText="Email"
    hintText="Enter your email address"
  />
</div>
```
