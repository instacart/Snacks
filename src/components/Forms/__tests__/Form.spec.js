import React         from 'react'
import renderer      from 'react-test-renderer'
import { StyleRoot } from 'radium'
import { mount }     from 'enzyme'
import { spy, stub } from 'sinon'
import Form          from '../Form'
import TextField     from '../TextField'
import Button        from '../../Buttons/Button'

it('renders Form with TextField correctly', () => {
  const tree = renderer.create(
    <StyleRoot>
      <Form>
        <TextField
          name="email"
          type="email"
          id="email_id"
          floatingLabelText="Email"
        />
      </Form>
    </StyleRoot>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

it('fires the onSubmit prop when valid', () => {
  const onSubmit = spy()
  const wrapper = mount(
    <Form
     onSubmit={onSubmit}
    >
      <div style={{width: '335px'}}>
        <div style={{marginBottom: '10px'}}>
          <TextField
            name="email"
            type="email"
            id="email_id"
            floatingLabelText="Email"
            value="test@test.com"
          />
        </div>

        <div style={{marginTop: '10px'}}>
          <Button type="submit">
           Submit
          </Button>
        </div>
      </div>
    </Form>
  )

  stub(wrapper.instance(), 'formIsValid').callsFake(() => (true))
  stub(wrapper.instance(), 'updateModel').callsFake(() => {})

  wrapper.find(Button).simulate('submit')
  expect(onSubmit.calledOnce).toBe(true)
})

it('wont fire the onSubmit prop when not valid', () => {
  const onSubmit = spy()
  const wrapper = mount(
    <Form
     onSubmit={onSubmit}
    >
      <div style={{width: '335px'}}>
        <div style={{marginBottom: '10px'}}>
          <TextField
            name="email"
            type="email"
            id="email_id"
            floatingLabelText="Email"
            value="test@test.com"
          />
        </div>

        <div style={{marginTop: '10px'}}>
          <Button type="submit">
           Submit
          </Button>
        </div>
      </div>
    </Form>
  )

  stub(wrapper.instance(), 'formIsValid').callsFake(() => (false))
  stub(wrapper.instance(), 'updateModel').callsFake(() => {})

  wrapper.find(Button).simulate('submit')
  expect(onSubmit.calledOnce).toBe(false)
})

it('wont fire the onSubmit prop when not valid', () => {
  const onSubmit = spy()
  const wrapper = mount(
    <Form
     onSubmit={onSubmit}
    >
      <div style={{width: '335px'}}>
        <div style={{marginBottom: '10px'}}>
          <TextField
            name="email"
            type="email"
            id="email_id"
            floatingLabelText="Email"
            value="test@test.com"
          />
        </div>

        <div style={{marginTop: '10px'}}>
          <Button type="submit">
           Submit
          </Button>
        </div>
      </div>
    </Form>
  )

  stub(wrapper.instance(), 'formIsValid').callsFake(() => (false))
  stub(wrapper.instance(), 'updateModel').callsFake(() => {})

  wrapper.find(Button).simulate('submit')
  expect(onSubmit.calledOnce).toBe(false)
})
