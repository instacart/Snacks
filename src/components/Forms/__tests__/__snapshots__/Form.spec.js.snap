// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Form with TextField correctly 1`] = `
.emotion-4 {
  cursor: auto;
  display: inline-block;
  position: relative;
  width: 343px;
}

.emotion-2 {
  border-radius: 4px;
  position: relative;
}

.emotion-0 {
  color: #757575;
  cursor: text;
  font-size: 16px;
  margin-top: 0;
  margin-right: 0;
  margin-bottom: 0;
  margin-left: 0;
  pointer-events: auto;
  position: absolute;
  -webkit-transform: scale(1) translate(8px,17px);
  -ms-transform: scale(1) translate(8px,17px);
  transform: scale(1) translate(8px,17px);
  -webkit-transform-origin: left top;
  -ms-transform-origin: left top;
  transform-origin: left top;
  -webkit-transition: all 450ms cubic-bezier(0.23,1,0.32,1) 0ms;
  transition: all 450ms cubic-bezier(0.23,1,0.32,1) 0ms;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  z-index: 1;
}

.emotion-1 {
  background-color: #FFF;
  border: solid 1px #BDBDBD;
  border-radius: 4px;
  box-sizing: border-box;
  color: #424242;
  font-size: 16px;
  height: 56px;
  margin-top: 0;
  margin-right: 0;
  margin-bottom: 0;
  margin-left: 0;
  padding-top: 25px;
  padding-left: 8px;
  padding-right: 8px;
  padding-bottom: 8px;
  position: relative;
  width: 100%;
  -webkit-opacity: 1px;
  -webkit-tap-highlight-color: rgba(0,0,0,0);
}

.emotion-3 {
  color: #B30029;
  font-size: 12px;
  margin-top: 2px;
  margin-left: 1px;
  opacity: 0;
  display: none;
  pointer-events: none;
  text-align: left;
  -webkit-transition: opacity 250ms cubic-bezier(0.23,1,0.32,1) 0ms;
  transition: opacity 250ms cubic-bezier(0.23,1,0.32,1) 0ms;
}

<form
  onSubmit={[Function]}
>
  <div
    className="emotion-4"
  >
    <div
      className="emotion-2"
    >
      <label
        className="emotion-0"
        htmlFor="email_id"
      >
        Email
      </label>
      <input
        aria-describedby=""
        aria-invalid={false}
        autoComplete="on"
        className="emotion-1"
        defaultValue={null}
        disabled={false}
        id="email_id"
        name="email"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        placeholder=""
        type="email"
      />
    </div>
    <div
      aria-atomic={true}
      aria-live="assertive"
      className="emotion-3"
      id="error_email_id"
    />
  </div>
</form>
`;
